### DISCUSSION TITLE
Firedrake update issues/Error in step computations

### DISCUSSION THREAD

**Discussion by mim-GK** [firedrakeproject/firedrake#3745] [Firedrake support]
Hi, I am currently working on a topology optimisation problem (code is shown)

```
from firedrake import *
from firedrake_adjoint import *
import matplotlib.pyplot as plt
import numpy as np
from firedrake import (
    inner,
    dx,
    ds,
    sym,
    Constant
)
from pyadjoint import Control, ReducedFunctional, MinimizationProblem, IPOPTSolver
from firedrake.adjoint import UFLInequalityConstraint

# Create a unit square mesh with 10x10 cells
mesh = UnitSquareMesh(300, 100)
X = SpatialCoordinate(mesh)
F = 2000 #####
traction = Constant([0, -F])
p = Constant(3)
eps = Constant(1e-3)
# Define a function space on the mesh
W = FunctionSpace(mesh, "CG", 1) ####
V = VectorFunctionSpace(mesh, "CG", 2)


# Structure material properties
# Elasticity parameters
E = Constant(1e5)
nu = Constant(0.3)
mu = (E / (2 * (1 + nu))) 
lmbda = (E * nu / ((1 + nu) * (1 - 2 * nu)))

# Define boundary conditions
# Fixed left edge (x = 0)
bc = DirichletBC(V, Constant((0, 0)), 1)  # Clamped at the left edge associated with vector values

def forward(rho):
    v = TestFunction(V)
    u = TrialFunction(V)
    # Define the traction force (applied in the middle of the right edge)

    
    L = inner(traction, v) * ds(2)

    # Variational problem setup
    a = 2*mu*inner(sym(grad(u)), sym(grad(v)))*dx + lmbda*div(u)*div(v)*dx

    # Define E_rho based on SIMP method
    E_rho = eps + (E - eps) * rho**p

    lmbda = nu * E_rho / ((1 + nu) * (1 - 2 * nu))
    mu = E_rho / (2 * (1 + nu))
    
    u = Function(V, name="Displacement")
    solve(a == L, u, bc, annotate=True)
    return u

# Set initial density and define the objective function
Vol = 0.5
rho = interpolate(Constant(Vol), W)
u = forward(rho)
J = assemble(inner(traction, u) * dx)####### could be 1 or dx
m = Control(rho)
Jhat = ReducedFunctional(J, m)
# Volume constraint
volume_constraint = UFLInequalityConstraint((Constant(Vol) - rho) * dx, m)

# Optimization setup
lb, ub = 0.0, 1.0
problem = MinimizationProblem(Jhat, bounds=(lb, ub), constraints=volume_constraint)
parameters = {"acceptable_tol": 1.0e-3, "maximum_iterations": 100}
solver = IPOPTSolver(problem, parameters=parameters)
rho_opt = solver.solve()

# PARAVIEW Results
results = File("newsolution.pvd")
results.parameters["flush_output"] = True
results.parameters["functions_share_mesh"] = True
results.write = (rho_opt, 0.)
results.write = (u, 0.)
```

where my code is resulting in an iteration but stops after the first one and results in an error as such:
[firedrake-update.log]
> FutureWarning: The firedrake_adjoint module is deprecated.
> 
> Instead, use the firedrake.adjoint module and explicitly start taping
> by calling continue_annotation().
>   warnings.warn("""The firedrake_adjoint module is deprecated.
> /Users/fire/firedrake/src/firedrake/firedrake/interpolation.py:385: FutureWarning: The use of `interpolate` to perform the numerical interpolation is deprecated.
> This feature will be removed very shortly.
> 
> Instead, import `interpolate` from the `firedrake.__future__` module to update
> the interpolation's behaviour to return the symbolic `ufl.Interpolate` object associated
> with this interpolation.
> 
> You can then assemble the resulting object to get the interpolated quantity
> of interest. For example,
> 
> ```
> from firedrake.__future__ import interpolate
> ...
> 
> assemble(interpolate(expr, V))
> ```
> 
> Alternatively, you can also perform other symbolic operations on the interpolation operator, such as taking
> the derivative, and then assemble the resulting form.
> 
>   warnings.warn("""The use of `interpolate` to perform the numerical interpolation is deprecated.
>  ** ERROR RETURN ** FROM DMUMPS INFO(1)=   -3
>  ** INFO(2)=              -1
> 
> List of options:
> 
>                                     Name   Value                # times used
>                    hessian_approximation = limited-memory            7
>                              print_level = 6                         2
> 
> ******************************************************************************
> This program contains Ipopt, a library for large-scale nonlinear optimization.
>  Ipopt is released as open source code under the Eclipse Public License (EPL).
>          For more information visit https://github.com/coin-or/Ipopt
> ******************************************************************************
> 
> This is Ipopt version 3.14.16, running with linear solver MUMPS 5.6.2.
> 
> Number of nonzeros in equality constraint Jacobian...:        0
> Number of nonzeros in inequality constraint Jacobian.:        0
> Number of nonzeros in Lagrangian Hessian.............:        0
> 
> Hessian approximation will be done in the space of all 36 x variables.
> 
> Scaling parameter for objective function = 1.000000e+00
> objective scaling factor = 1
> No x scaling provided
> No c scaling provided
> No d scaling provided
> Initial values of x sufficiently inside the bounds.
> Initial values of s sufficiently inside the bounds.
> Total number of variables............................:       36
>                      variables with only lower bounds:        0
>                 variables with lower and upper bounds:        0
>                      variables with only upper bounds:        0
> Total number of equality constraints.................:        0
> Total number of inequality constraints...............:        0
>         inequality constraints with only lower bounds:        0
>    inequality constraints with lower and upper bounds:        0
>         inequality constraints with only upper bounds:        0
> 
> 
> **************************************************
> *** Update HessianMatrix for Iteration 0:
> **************************************************
> 
> Limited-Memory approximation started; store data at current iterate.
> 
> 
> **************************************************
> *** Summary of Iteration: 0:
> **************************************************
> 
> iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
>    0  3.4740741e-04 0.00e+00 3.27e-03   0.0 0.00e+00    -  0.00e+00 0.00e+00   0 
> 
> **************************************************
> *** Beginning Iteration 0 from the following point:
> **************************************************
> 
> Current barrier parameter mu = 1.0000000000000000e+00
> Current fraction-to-the-boundary parameter tau = 0.0000000000000000e+00
> 
> ||curr_x||_inf   = 5.0000000000000000e-01
> ||curr_s||_inf   = 0.0000000000000000e+00
> ||curr_y_c||_inf = 0.0000000000000000e+00
> ||curr_y_d||_inf = 0.0000000000000000e+00
> ||curr_z_L||_inf = 0.0000000000000000e+00
> ||curr_z_U||_inf = 0.0000000000000000e+00
> ||curr_v_L||_inf = 0.0000000000000000e+00
> ||curr_v_U||_inf = 0.0000000000000000e+00
> 
> 
> ***Current NLP Values for Iteration 0:
> 
>                                    (scaled)                 (unscaled)
> Objective...............:   3.4740740740740750e-04    3.4740740740740750e-04
> Dual infeasibility......:   3.2717037037037055e-03    3.2717037037037055e-03
> Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
> Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
> Overall NLP error.......:   3.2717037037037055e-03    3.2717037037037055e-03
> 
> 
> 
> 
> **************************************************
> *** Update Barrier Parameter for Iteration 0:
> **************************************************
> 
> Setting mu_max to 0.000000e+00.
> Barrier Parameter: 1.000000e-11
> 
> **************************************************
> *** Solving the Primal Dual System for Iteration 0:
> **************************************************
> 
>  ** ERROR RETURN ** FROM DMUMPS INFO(1)=   -3
>  ** INFO(2)=               1
> MUMPS used permuting_scaling 0 and pivot_order 0.
>            scaling will be 77.
> Error=-3 returned from MUMPS in Factorization.
> Factorization failed with retval = 4
> LowRankAugSystemSolver: AugSystemSolver returned retval = 4 for right hand side.
> *** Step could not be computed in iteration 0!
> ERROR: Problem in step computation, but emergency mode cannot be activated.
> 
> Number of Iterations....: 0
> 
>                                    (scaled)                 (unscaled)
> Objective...............:   3.4740740740740750e-04    3.4740740740740750e-04
> Dual infeasibility......:   3.2717037037037055e-03    3.2717037037037055e-03
> Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
> Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
> Complementarity.........:   0.0000000000000000e+00    0.0000000000000000e+00
> Overall NLP error.......:   3.2717037037037055e-03    3.2717037037037055e-03
> 
> 
> Number of objective function evaluations             = 1
> Number of objective gradient evaluations             = 1
> Number of equality constraint evaluations            = 0
> Number of inequality constraint evaluations          = 0
> Number of equality constraint Jacobian evaluations   = 0
> Number of inequality constraint Jacobian evaluations = 0
> Number of Lagrangian Hessian evaluations             = 0
> Total seconds in IPOPT                               = 4.648
> 
> EXIT: Error in step computation!
>  ** ERROR RETURN ** FROM DMUMPS INFO(1)=   -3
>  ** INFO(2)=              -2

I tried updating firedrake to ensure the installation is up-to-date and properly configured. However, this resulted in the following and I have attached the Update log along with it
[firedrake-update.log](https://github.com/user-attachments/files/16741997/firedrake-update.log)
:

> /Users/fire/firedrake/bin/firedrake-update:20: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
>   from pkg_resources.extern.packaging.version import Version, InvalidVersion
> Running /Users/fire/firedrake/bin/firedrake-update
> BLAS config is {'BLAS': '/Users/fire/firedrake/src/petsc/default', 'OPENBLAS': '/Users/fire/firedrake/src/petsc/default', 'LDFLAGS': '-L/usr/local/opt/gcc/lib/gcc/14'}
> Configuration saved to /Users/fire/firedrake/.configuration.json
> Updating the git repository for firedrake
> Creating firedrake-update script.
> Running ../bin/firedrake-update --no-update-script
> Installing command line tools...
> Checking brew doctor...
> Installing required packages via homebrew...
> Required package 'gcc' is already installed via Homebrew.
> Required package 'autoconf' is already installed via Homebrew.
> Required package 'make' is already installed via Homebrew.
> Required package 'automake' is already installed via Homebrew.
> Required package 'cmake' is already installed via Homebrew.
> Installing required package 'ninja' via Homebrew.
> Required package 'libtool' is already installed via Homebrew.
> Required package 'boost' is already installed via Homebrew.
> BLAS config is {'BLAS': '/Users/fire/firedrake/src/petsc/default', 'OPENBLAS': '/Users/fire/firedrake/src/petsc/default', 'LDFLAGS': '-L/usr/local/opt/gcc/lib/gcc/14'}
> Configuration saved to /Users/fire/firedrake/.configuration.json
> Finding dependencies of PyOP2
> 
> Finding dependencies of firedrake
> 
> Updating the git repository for petsc
> Removing existing petsc installations
> 
> Updating the git repository for loopy
> Updating the git repository for ufl
> Updating the git repository for fiat
> Updating the git repository for FInAT
> Updating the git repository for tsfc
> Updating the git repository for PyOP2
> Updating the git repository for pyadjoint
> Updating the git repository for pytest-mpi
> Updating the git repository for firedrake
> Cleaning petsc/
> Installing petsc/
> Depending on your platform, PETSc may take an hour or more to build!
> Traceback (most recent call last):
>   File "/Users/fire/firedrake/src/../bin/firedrake-update", line 1960, in <module>
>     install("petsc/")
>   File "/Users/fire/firedrake/src/../bin/firedrake-update", line 1043, in install
>     build_and_install_petsc()
>   File "/Users/fire/firedrake/src/../bin/firedrake-update", line 1155, in build_and_install_petsc
>     check_call([python, "./configure", "PETSC_DIR={}".format(petsc_dir), "PETSC_ARCH={}".format(petsc_arch)] + petsc_options)
>   File "/Users/fire/firedrake/src/../bin/firedrake-update", line 672, in check_call
>     log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/Cellar/python@3.11/3.11.9/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 466, in check_output
>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/Cellar/python@3.11/3.11.9/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 571, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['/Users/fire/firedrake/bin/python', './configure', 'PETSC_DIR=/Users/fire/firedrake/src/petsc', 'PETSC_ARCH=default', '--download-mumps', '--download-netcdf', '--download-scalapack', '--with-fortran-bindings=0', '--with-c2html=0', '--with-zlib', '--download-hdf5', '--download-hwloc', '--download-pnetcdf', '--download-suitesparse', '--with-shared-libraries=1', '--with-x=0', '--download-pastix', '--CFLAGS=-Wno-implicit-function-declaration', '--download-superlu_dist', '--with-debugging=0', "--download-openblas-make-options='USE_THREAD=0 USE_LOCKING=1 USE_OPENMP=0'", '--download-mpich', '--download-ptscotch', '--download-hwloc-configure-arguments=--disable-opencl', '--download-hypre', '--download-mpich-configure-arguments=--disable-opencl', '--download-openblas', '--download-bison', '--download-metis', '--LDFLAGS=-Wl,-ld_classic']' returned non-zero exit status 1.
> 
> 
> Update log saved in /Users/fire/firedrake/firedrake-update.log

As a result, my code is not working at all and is stating the following:

> Traceback (most recent call last):
>   File "/Users/fire/canweend.py", line 1, in <module>
>     from firedrake import *
>   File "/Users/fire/firedrake/src/firedrake/firedrake/__init__.py", line 1, in <module>
>     import firedrake_configuration
>   File "/Users/fire/firedrake/src/firedrake/firedrake_configuration/__init__.py", line 37, in <module>
>     options["with_parmetis"] = "parmetis" in petsc_packages()
>                                              ^^^^^^^^^^^^^^^^
>   File "/Users/fire/firedrake/src/firedrake/firedrake_configuration/__init__.py", line 31, in petsc_packages
>     with open(os.path.join(conf["PETSC_DIR"], conf["PETSC_ARCH"], "include", "petscconf.h"), "r") as f:
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> FileNotFoundError: [Errno 2] No such file or directory: '/Users/fire/firedrake/src/petsc/default/include/petscconf.h'
>

**Comment by Ig-dolci** [comment]
Unfortunately, there have been issues with mumps when using parallel solvers on macOS. Could you please modify your solver by setting `solve(a == L, u, bc, solver_parameters={"pc_factor_mat_solver_type": "superlu_dist"})`? Note that you do not need to include `annotate=True`. To ensure your forward solver is annotated for the automated gradient computations, move from `from firedrake_adjoint import *` to `from firedrake.adjoint import *` and `continue_annotation()` right before your forward solver as shown in this [notebook](https://nbviewer.org/github/firedrakeproject/firedrake/blob/master/docs/notebooks/06-pde-constrained-optimisation.ipynb).

**Reply by mim-GK** [reply]
@Ig-dolci my petsc files has disappeared completely from my src file when I tried firedrake-update and now none of my firedrake codes are running, what could I do?

**Reply by Ig-dolci** [reply]
Try `firedrake-update --rebuild`. If it does not work, reinstalling Firedrake will be necessary.

**Reply by mim-GK** [reply]
Thank you @Ig-dolci I managed to re-install and tried running with the recommended edits, but I am receiving these errors for the following updated code even though I have installed cyipopt:

> UserWarning: Unable to import recommended hash 'siphash24.siphash13', falling back to 'hashlib.sha256'. Run 'python3 -m pip install siphash24' to install the recommended hash.
>   warn("Unable to import recommended hash 'siphash24.siphash13', "
> firedrake:WARNING OMP_NUM_THREADS is not set or is set to a value greater than 1, we suggest setting OMP_NUM_THREADS=1 to improve performance
> /Users/fire/firedrake/src/firedrake/firedrake/interpolation.py:385: FutureWarning: The use of `interpolate` to perform the numerical interpolation is deprecated.
> This feature will be removed very shortly.
> 
> Instead, import `interpolate` from the `firedrake.__future__` module to update
> the interpolation's behaviour to return the symbolic `ufl.Interpolate` object associated
> with this interpolation.
> 
> You can then assemble the resulting object to get the interpolated quantity
> of interest. For example,
> 
> ```
> from firedrake.__future__ import interpolate
> ...
> 
> assemble(interpolate(expr, V))
> ```
> 
> Alternatively, you can also perform other symbolic operations on the interpolation operator, such as taking
> the derivative, and then assemble the resulting form.
> 
>   warnings.warn("""The use of `interpolate` to perform the numerical interpolation is deprecated.
> 
> A module that was compiled using NumPy 1.x cannot be run in
> NumPy 2.1.0 as it may crash. To support both 1.x and 2.x
> versions of NumPy, modules must be compiled with NumPy 2.0.
> Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.
> 
> If you are a user of the module, the easiest solution will be to
> downgrade to 'numpy<2' or try to upgrade the affected module.
> We expect that some modules will need time to support NumPy 2.
> 
> Traceback (most recent call last):  File "/Users/fire/TopOpt.py", line 75, in <module>
>     solver = IPOPTSolver(problem, parameters=parameters)
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/optimization/ipopt_solver.py", line 31, in __init__
>     self.__build_ipopt_problem()
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/optimization/ipopt_solver.py", line 37, in __build_ipopt_problem
>     from pyadjoint.ipopt import cyipopt
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/ipopt.py", line 11, in <module>
>     import cyipopt  # noqa: F401
>   File "/Users/fire/firedrake/lib/python3.11/site-packages/cyipopt/__init__.py", line 12, in <module>
>     from ipopt_wrapper import *
> Traceback (most recent call last):
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/ipopt.py", line 11, in <module>
>     import cyipopt  # noqa: F401
>     ^^^^^^^^^^^^^^
>   File "/Users/fire/firedrake/lib/python3.11/site-packages/cyipopt/__init__.py", line 12, in <module>
>     from ipopt_wrapper import *
>   File "cyipopt/cython/ipopt_wrapper.pyx", line 1, in init ipopt_wrapper
> ImportError: numpy.core.multiarray failed to import (auto-generated because you didn't call 'numpy.import_array()' after cimporting numpy; use '<void>numpy._import_array' to disable if you are certain you don't need it).
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/Users/fire/TopOpt.py", line 75, in <module>
>     solver = IPOPTSolver(problem, parameters=parameters)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/optimization/ipopt_solver.py", line 31, in __init__
>     self.__build_ipopt_problem()
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/optimization/ipopt_solver.py", line 37, in __build_ipopt_problem
>     from pyadjoint.ipopt import cyipopt
>   File "/Users/fire/firedrake/src/pyadjoint/pyadjoint/ipopt.py", line 13, in <module>
>     raise ImportError("You need to install cyipopt. It is recommended to install IPOPT with HSL support!")
> ImportError: You need to install cyipopt. It is recommended to install IPOPT with HSL support!

`from firedrake import *
from firedrake.adjoint import *
import matplotlib.pyplot as plt
import numpy as np
from firedrake import (
    inner,
    dx,
    ds,
    sym,
    Constant
)
from pyadjoint import Control, ReducedFunctional, MinimizationProblem, IPOPTSolver
from firedrake.adjoint import UFLInequalityConstraint

continue_annotation()
# Create a unit square mesh with 10x10 cells
mesh = UnitSquareMesh(300, 100)
X = SpatialCoordinate(mesh)
F = 2000 #####
traction = Constant([0, -F])
p = Constant(3)
eps = Constant(1e-3)
# Define a function space on the mesh
W = FunctionSpace(mesh, "CG", 1) ####
V = VectorFunctionSpace(mesh, "CG", 2)


# Structure material properties
# Elasticity parameters
E = Constant(1e5)
nu = Constant(0.3)
mu = (E / (2 * (1 + nu))) 
lmbda = (E * nu / ((1 + nu) * (1 - 2 * nu)))

# Define boundary conditions
# Fixed left edge (x = 0)
bc = DirichletBC(V, Constant((0, 0)), 1)  # Clamped at the left edge associated with vector values

def forward(rho):
    v = TestFunction(V)
    u = TrialFunction(V)
    # Define the traction force (applied in the middle of the right edge)

    

    # Define E_rho based on SIMP method
    E_rho = eps + (E - eps) * rho**p

    lmbda = nu * E_rho / ((1 + nu) * (1 - 2 * nu))
    mu = E_rho / (2 * (1 + nu))

    L = inner(traction, v) * ds(2)

    # Variational problem setup
    a = 2*mu*inner(sym(grad(u)), sym(grad(v)))*dx + lmbda*div(u)*div(v)*dx
    
    u = Function(V, name="Displacement")
    solve(a == L, u, bc, solver_parameters={"pc_factor_mat_solver_type": "superlu_dist"})
    return u

# Set initial density and define the objective function
Vol = 0.5
rho = interpolate(Constant(Vol), W)
u = forward(rho)
J = assemble(inner(traction, u) * dx)####### could be 1 or dx
m = Control(rho)
Jhat = ReducedFunctional(J, m)
# Volume constraint
volume_constraint = UFLInequalityConstraint((Constant(Vol) - rho) * dx, m)

# Optimization setup
lb, ub = 0.0, 1.0
problem = MinimizationProblem(Jhat, bounds=(lb, ub), constraints=volume_constraint)
parameters = {"acceptable_tol": 1.0e-3, "maximum_iterations": 100}
solver = IPOPTSolver(problem, parameters=parameters)
rho_opt = solver.solve()

# PARAVIEW Results
results = File("newsolution.pvd")
results.parameters["flush_output"] = True
results.parameters["functions_share_mesh"] = True
results.write = (rho_opt, 0.)
results.write = (u, 0.)`

**Reply by Ig-dolci** [reply]
Have you installed `cyipopt` in the Firedrake virtual environment?

**Reply by mim-GK** [reply]
@Ig-dolci Yes I installed cyipopt with the following and this resulted:

> pip install cyipopt
> 
> Collecting cyipopt
>   Using cached cyipopt-1.4.1-cp311-cp311-macosx_10_9_universal2.whl
> Requirement already satisfied: numpy>=1.21.5 in ./firedrake/lib/python3.11/site-packages (from cyipopt) (2.1.0)
> Installing collected packages: cyipopt
> Successfully installed cyipopt-1.4.1

**Reply by connorjward** [reply]
This seems to be an issue with cyipopt. I expect that you could reproduce the issue just by running (with the venv activated) `python -c "import cyipopt"`. I suggest contacting them to report the issue and find a fix.

**Reply by mim-GK** [reply]
Yes, as expected running `python -c "import cyipopt"` repeats the same errors unfortonutaly. In relation to my code are there any alternatives that would be compatible with the NumPy version within firedrake?

**Reply by Ig-dolci** [reply]
Sorry. Could you please clarify "NumPy version within firedrake"? Do you mean another package that solves optimisation problems with inequality constraints?

**Reply by mim-GK** [reply]
Yes @Ig-dolci, are there any alternative ipopt wrappers that I could use with this problem within firedrake or it has to be cyipopt since the previous outputs suggested only cyipopt?

**Reply by Ig-dolci** [reply]
I saw some users using the pyrol package. It looks like this [pyrol repository](https://github.com/angus-g/pyrol) is maintainable. Take a look at the examples there and how to install it.
